TMU/SL/BCM.c

```diff
@@ -0,0 +1,693 @@
+
+/* RX Dispatcher */
+void BCM_Rx_Dispatcher(void)
+{
+	if(g_BCM_EXcfg.FSM_State == ReceiveComplete_State)
+	{	
+		if(g_BCM_EXcfg.CheckSum == g_CS_Calculated)
+		{
+		
+			g_BCM_EXcfg.Count = ZERO;
+			g_CS_Calculated = ZERO;
+			g_BCM_EXcfg.CheckSum = ZERO;
+			g_BCM_EXcfg.FSM_State = IDLE_State;
+			g_BCM_EXcfg.BCM_notify_cbf(E_OK);
+		}
+	}
+}
+
+/* TX Dispatcher */
+void BCM_Tx_Dispatcher(void)
+{	
+	switch(g_BCM_EXcfg.FSM_State)
+	{
+		case IDLE_State :
+			break;
+		case SendingByte_State :
+			break;
+		case SendComplete_State :
+		if(g_BCM_EXcfg.Count == 1)
+		{
+			switch(g_BCM_EXcfg.Protocol){
+				case UART_Protocol :
+				UART_Write(g_BCM_EXcfg.Buf_Len);
+				break;
+				case  SPI_Protocol :
+				_SPISend(g_BCM_EXcfg.Buf_Len);
+				break;
+			}
+			g_BCM_EXcfg.FSM_State = SendingByte_State ;
+			
+		}
+		else if (g_BCM_EXcfg.Count <= g_BCM_EXcfg.Buf_Len+1)
+		{
+			g_BCM_EXcfg.CheckSum +=g_BCM_EXcfg.Buffer[g_BCM_EXcfg.Count-2];
+			
+			switch(g_BCM_EXcfg.Protocol){
+				case UART_Protocol :
+				UART_Write(g_BCM_EXcfg.Buffer[g_BCM_EXcfg.Count-2]);
+				break;
+				case  SPI_Protocol :
+				_SPISend(g_BCM_EXcfg.Buffer[g_BCM_EXcfg.Count-2]);	
+				break;
+			}
+			g_BCM_EXcfg.FSM_State = SendingByte_State ;
+
+			}else if (g_BCM_EXcfg.Count == g_BCM_EXcfg.Buf_Len+2){
+			switch(g_BCM_EXcfg.Protocol){
+				
+				case UART_Protocol :
+				UART_Write(g_BCM_EXcfg.CheckSum);
+				break;
+				case  SPI_Protocol :
+				_SPISend(g_BCM_EXcfg.CheckSum);
+				break;
+			}
+			g_BCM_EXcfg.FSM_State = SendingByte_State ;
+			
+			}else{
+				BCM_TxUnlock(&g_BCM_EXcfg);
+				g_BCM_EXcfg.BCM_notify_cbf(10);
+			g_BCM_EXcfg.FSM_State = IDLE_State ;
+		}
+		break;
+		case OFF_State :
+		break;
+	}	
+}
+/*------------------------------------*/
+
+/* BCM Initialization Routine */
+
+/**
+ * Input: Pointer to a structure contains the information needed to initialize the BCM 
+ * Output:
+ * In/Out:			
+ * Return: The error status of the function.			
+ * Description: Initiates the Basic Communication Module
+ * 							
+ */
+
+
+ERROR_STATUS BCM_Init(BCM_cfg_s* a_BCM)
+{
+	ERROR_STATUS errorStatus = BCM_ERROR + E_NOK;
+	UART_cfg  a_BCM_UART;
+	SPI_cfg_s a_BCM_SPI; 
+	
+	if (g_BCM_EXcfg.FSM_State==IDLE_State)
+	{
+		 errorStatus =  BCM_ERROR + ALREADY_INIT ;
+		 return errorStatus ;
+	}
+	
+	
+	/*-------------[ Check BCM's Pointer Validity ]-------------*/
+	if(a_BCM != NULL)
+	{
+		/*--------[ Add The BCM CFG To The Working List ]--------*/
+		g_BCM_EXcfg.BCM_CH_ID = a_BCM->BCM_CH_ID;
+		g_BCM_EXcfg.Mode      = a_BCM->Mode;
+		g_BCM_EXcfg.Protocol  = a_BCM->Protocol;
+		g_BCM_EXcfg.FSM_State = IDLE_State;
+		g_BCM_EXcfg.Count     = ZERO;
+		g_BCM_EXcfg.CheckSum  = ZERO;
+		g_BCM_EXcfg.Buf_Len   = ZERO;
+		g_BCM_EXcfg.MSG_Len   = ZERO;
+		
+		/*--------[ Check The BCM HW Communication Protocol ]--------*/
+		switch(g_BCM_EXcfg.Protocol)
+		{
+			/*----------------------[ UART CASE START ]---------------------*/
+			case UART_Protocol:
```

Comment:
I like your code style but I have a comment of usage the SPI and UART as this method let you have duplicated code ... Is there a better way to select between both even in runtime/ in config ?